### PR TITLE
[gdal] Add poppler

### PR DESCRIPTION
**Pull request by m-kuhn**
**Describe the pull request**

- #### What does your PR fix?  
Adds poppler / pdf support

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  linux, macs (windows to be done)

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

### CODE REVIEW

**Comment by m-kuhn** [comment]
Todo : update vcpkg-cmake-wrapper and add poppler as linker dependency

**Comment by dg0yt** [comment]
> Todo : update vcpkg-cmake-wrapper and add poppler as linker dependency

And put it behind a feature. And handle windows (nmake). And wait for the 3.4.1 PR to be merged first.

**Comment by m-kuhn** [comment]
- feature added
- windows added

I'd be happy if I could get guidance on how to best update `vcpkg-cmake-wrapper.cmake`, poppler does not provide any cmake config, should I play with FindPkgConfig or is there a better way to go about that?

**Comment by dg0yt** [comment]
> I'd be happy if I could get guidance on how to best update `vcpkg-cmake-wrapper.cmake`, poppler does not provide any cmake config, should I play with FindPkgConfig or is there a better way to go about that?

The missing cmake stuff is the reason why I didn't already add it. `FindPkgConfig` only helps when poppler`s pc file available, complete and valid for all platforms. If this is the case, pkgconfig can already be used to configure nmake. 

I completely rewrote nmake configuration in #22370. I don't want this to be delayed by new merge conflicts.

Sooner or later, we will need exported cmake config from poppler, at least "unofficial".

**Comment by m-kuhn** [comment]
Thanks.

I'll leave this in draft until #22370 is merged. Nice improvements there.
Is there a best practices example of how to create an unofficial exported cmake config? There seem to be a couple of ports already doing this.

**Comment by dg0yt** [comment]
> Is there a best practices example of how to create an unofficial exported cmake config? There seem to be a couple of ports already doing this.

I can't claim "best practice", but libmariadb might serve as an example. Basically there are two aspects.
- call the export in CMake (easy)
- inject the needed `find_dependency(...)` calls to the config (advanced).

The `find_dependency` stuff is usually done in the main config file (`pkg-config.cmake` or `PkgConfig.cmake`),
and then the primary exported file (e.g. `pkg-targets.cmake`) is chainloaded.

Many ports have the config template as a plain file, and configure it on install as needed by features.
In libmariadb, writing the config file is patched into build system. 
Best practice is to minimize patching, but sometimes it is easier if you can derived variables in the build system.

File: ports/poppler/export-cmake-targets.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
++install(EXPORT unofficial-poppler-targets
++    NAMESPACE unofficial::
++    DESTINATION share/unofficial-poppler
++)
++file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/unofficial-poppler-config.cmake.in"
++[[include(CMakeFindDependencyMacro)
++find_dependency(Freetype)
++find_dependency(LCMS2)
++if(@ENABLE_CURL@)
++    find_dependency(OpenSSL)
```
Comment (dg0yt) [inline ports/poppler/export-cmake-targets.patch:20]: If the condition is `@ENABLE_CURL@`, shouldn't it be `find_dependency(CURL)`?

File: ports/poppler/export-cmake-targets.patch [hunk 2]
```diff
@@ -0,0 +1,31 @@
+diff --color -ur /home/mkuhn/dev/vcpkg/buildtrees/poppler/src/er-20.12.1-0f846aceea.clean/CMakeLists.txt /home/mkuhn/dev/vcpkg/buildtrees/poppler/src/er-20.12.1-0f846aceea/CMakeLists.txt
+--- /home/mkuhn/dev/vcpkg/buildtrees/poppler/src/er-20.12.1-0f846aceea.clean/CMakeLists.txt	2022-01-19 18:49:42.309363752 +0100
++++ /home/mkuhn/dev/vcpkg/buildtrees/poppler/src/er-20.12.1-0f846aceea/CMakeLists.txt	2022-01-19 20:34:22.171542334 +0100
+@@ -578,7 +578,26 @@
+ if(CMAKE_USE_PTHREADS_INIT)
+    target_link_libraries(poppler LINK_PRIVATE Threads::Threads)
+ endif()
+-install(TARGETS poppler RUNTIME DESTINATION bin LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR} ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
++install(TARGETS poppler EXPORT unofficial-poppler-targets RUNTIME DESTINATION bin LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR} ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
```
Comment (dg0yt) [inline ports/poppler/export-cmake-targets.patch:9]: I would prefer double quotes around `${CMAKE_INSTALL_LIBDIR}` and similar expression.

File: ports/poppler/export-cmake-targets.patch [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
++if(@ENABLE_CURL@)
++    find_dependency(OpenSSL)
++endif()
++if(@ENABLE_ZLIB@)
++    find_dependency(ZLIB)
++endif()
++include("${CMAKE_CURRENT_LIST_DIR}/unofficial-poppler-targets.cmake")
++]])
++configure_file("${CMAKE_CURRENT_BINARY_DIR}/unofficial-poppler-config.cmake.in" "${CMAKE_CURRENT_BINARY_DIR}/unofficial-poppler-config.cmake" @ONLY)
++install(FILES ${CMAKE_CURRENT_BINARY_DIR}/unofficial-poppler-config.cmake DESTINATION share/unofficial-poppler)
```
Comment (dg0yt) [inline ports/poppler/export-cmake-targets.patch:28]: Double quotes also here.

File: ports/poppler/portfile.cmake [hunk 4]
```diff
@@ -15,12 +15,13 @@ get_filename_component(GPERF_PATH ${GPERF} DIRECTORY)
 vcpkg_add_to_path(${GPERF_PATH})
 
 vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+  FEATURES
```
Comment (dg0yt) [inline ports/poppler/portfile.cmake:18]: I assume other reviewer will ask for different indentation of this and the following lines.

**Review by dg0yt — CHANGES_REQUESTED** [review]
I'm testing it locally and will send you a PR with additional poppler changes.

**Comment by dg0yt** [comment]
I think a should create a separate PR for updating poppler. There are a few more issues.

**Comment by m-kuhn** [comment]
Thank you for the feedback. I will go over it in the next hours.

**Comment by m-kuhn** [comment]
I'm taking a look at the patch which isn't applied at the moment

**Comment by dg0yt** [comment]
> I'm taking a look at the patch which isn't applied at the moment

This didn't stop my local tests ;-) 
Poppler dependencies in vcpkg are completely broken. It requests some ports it doesn't use, and it uses some ports which it doesn't request. We must solve this first, or the exported config will not be reliable: garbage in, garbage out.

**Comment by m-kuhn** [comment]
I noticed the optional deps in the CMakeLists.txt, they will be hard to handle without patching poppler itself -- or making them all mandatory -- if I'm not mistaken.

**Comment by dg0yt** [comment]
More on poppler:
- We need to export the public include directory.
- The poppler frontends come as separate libraries. We might also need to export `poppler_cpp`.
- Dependency control should be doable without patching. The optional stuff is controlled via `WITH_`. This is hidden in `macro_optional_find_package`.

My work-in-progress relies on my vcpg-cmake-validate PR in order to validate the exported config right on building poppler. I will create a poppler PR without this dependency when it passed that check...

**Comment by dg0yt** [comment]
Now I identified a bug in the Fontconfig wrapper (losing freetype and expat dependencies). Current work list:
- [ ] Fix fontconfig
- [ ] Do unofficial::poppler
- [ ] Add poppler to gdal

**Comment by m-kuhn** [comment]
Please let me know if I can help, I have currently put this on hold to avoid conflicts

**Comment by m-kuhn** [comment]
Superseded by https://github.com/microsoft/vcpkg/pull/22720

**Pull request closed without merging**